packages/runner/injection/patches/cookies.ts

```diff
@@ -0,0 +1,127 @@
+      sameSite: 'lax',
+      secure: false,
+    }
+  })
+}
+
+const sendCookieToServer = (cookie: AutomationCookie) => {
+  window.top!.postMessage({
+    event: 'cross:origin:aut:set:cookie',
+    data: {
+      cookie,
+      url: location.href,
+      // url will always match the cookie domain, so strict context tells
+      // tough-cookie to allow it to be set
+      sameSiteContext: 'strict',
+    },
+  }, '*')
+}
+
+// document.cookie monkey-patching
+// -------------------------------
+// We monkey-patch document.cookie when in a cross-origin injection, because
+// document.cookie runs into cross-origin restrictions when the AUT is on
+// a different origin than top. The goal is to make it act like it would
+// if the user's app was run in top.
+export const patchDocumentCookie = (requestCookies: AutomationCookie[]) => {
+  const url = location.href
+  const domain = location.hostname
+  const cookieJar = new CookieJar()
+  const existingCookies = parseDocumentCookieString(document.cookie)
+
+  const getDocumentCookieValue = () => {
+    return cookieJar.getCookies(url, undefined).map((cookie: ToughCookie) => {
+      return `${cookie.key}=${cookie.value}`
+    }).join('; ')
+  }
+
+  const addCookies = (cookies: AutomationCookie[]) => {
+    cookies.forEach((cookie) => {
+      cookieJar.setCookie(automationCookieToToughCookie(cookie, domain), url, undefined)
+    })
+  }
+
+  // requestCookies are ones included with the page request that's now being
+  // injected into. they're captured by the proxy and included statically in
+  // the injection so they can be added here and available before page load
+  addCookies(existingCookies.concat(requestCookies))
+
+  Object.defineProperty(window.document, 'cookie', {
+    get () {
+      return getDocumentCookieValue()
+    },
+
+    set (newValue: any) {
+      const stringValue = `${newValue}`
+      const parsedCookie = CookieJar.parse(stringValue)
+
+      // if result is undefined, it was invalid and couldn't be parsed
+      if (!parsedCookie) return getDocumentCookieValue()
+
+      // we should be able to pass in parsedCookie here instead of the string
+      // value, but tough-cookie doesn't recognize it using an instanceof
+      // check and throws an error. because we can't, we have to massage
+      // some of the properties below to be correct
+      const cookie = cookieJar.setCookie(stringValue, url, undefined)!
+
+      cookie.sameSite = parsedCookie.sameSite
+
+      if (!parsedCookie.path) {
+        cookie.path = '/'
+      }
+
+      // send the cookie to the server so it can be set in the browser via
+      // automation and in our server-side cookie jar so it's available
+      // to subsequent injections
+      sendCookieToServer(toughCookieToAutomationCookie(cookie, domain))
+
+      return getDocumentCookieValue()
+    },
+  })
+
+  const reset = () => {
+    cookieJar.removeAllCookies()
+  }
+
+  const bindCypressListeners = (Cypress: Cypress.Cypress) => {
```

Comment:
Currently that fails with `The command was expected to run against origin http://localhost:3500 but the application is at origin http://www.foobar.com:3500`. I think it makes sense that the user should put it inside the origin callback in that case and then the listeners will be set up as well.

Reply:
I think we may want to revisit throwing that error:
>The command was expected to run against origin http://localhost:3500 but the application is at origin http://www.foobar.com:3500

I added the limitation that cy.cookies could only be run when the aut was same-site as top, I don't think thats correct anymore since we will be allowing cy.cookies to be set on any domain (i think it can be done now?) and we should probably think about changing this behaviour in #24264